File: K_Sum_Paths.py
class Solution:
    def solve(self, root, k, total, sum, ending, mod):
        if root is None:
           return
        sum += root.data
        total[0] = (total[0] + ending.get(sum - k, 0)) % mod
        ending[sum] = ending.get(sum, 0) + 1
        Solution().solve(root.left, k, total, sum, ending, mod)
        Solution().solve(root.right, k, total, sum, ending, mod)
        ending[sum] -= 1
    def sumK(self,root,k):
        total = [0]
        mod = int(1e9) + 7
        ending = {0: 1}
        Solution().solve(root, k, total, 0, ending, mod)
        return total[0]

File: K_Sum_Paths.cpp
class Solution{
  public:
    const int mod = 1e9 + 7;
    int total = 0;
    void solve(Node *root, int k, int sum, unordered_map<int, int> &ending){
        if(!root){
            return;
        }
        sum += root -> data;
        total = (total + ending[sum - k])%mod;
        ending[sum]++;
        solve(root -> left, k, sum, ending);
        solve(root -> right, k, sum, ending);
        ending[sum]--;
    }
    int sumK(Node *root, int k){
        unordered_map<int, int> ending;
        ending[0] = 1;
        solve(root, k, 0, ending);
        return total;
    }
};
